### PR TITLE
fix: fix mobile scanner startup bug on web

### LINKED ISSUE #622: Exception with Flutter 3.10
with flutter 3.10 i get this exception on startup:

TypeError: Cannot set properties of null (setting 'AbstractExpandedDecoder')
https://unpkg.com/@zxing/library@0.19.1 15:288561  <fn>
https://unpkg.com/@zxing/library@0.19.1 1:69       <fn>
https://unpkg.com/@zxing/library@0.19.1 1:206      <fn>

reproduce:
Add mobile_scanner to pubspec.yaml

### LINKED ISSUE #631: Flutter 3.10 (tracker issue)
### What

- [x] #622 
- [x] #637 

### Part of
- #260

### LINKED ISSUE #632: Flutter 3.10 Permission Denied in flutter web...
My Flutter app runs on the web. I'm getting permission denied on Android using the Chrome browser. README doesn't have any android permission info or anything I need to include in index.html? How do I get past this on Flutter web?

image

Version: mobile_scanner: ^3.2.0

```dart
late final MobileScannerController _controller = MobileScannerController();

MobileScanner(
    controller: _controller,
    fit: BoxFit.contain,
    errorBuilder: (
      BuildContext context,
      MobileScannerException error,
      Widget? child,
    ) {
      return ScannerErrorWidget(error: error);
    },
    onDetect: widget.onDetect,
  )

class ScannerErrorWidget extends StatelessWidget {
  const ScannerErrorWidget({super.key, required this.error});

  final MobileScannerException error;

  @override
  Widget build(BuildContext context) {
    String errorMessage;

    switch (error.errorCode) {
      case MobileScannerErrorCode.controllerUninitialized:
        errorMessage = 'Controller not ready.';
        break;
      case MobileScannerErrorCode.permissionDenied:
        errorMessage = 'Permission denied';
        break;
      default:
        errorMessage = 'Generic Error';
        break;
    }

    return ColoredBox(
      color: Colors.black,
      child: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            const Padding(
              padding: EdgeInsets.only(bottom: 16),
              child: Icon(Icons.error, color: Colors.white),
            ),
            Text(
              errorMessage,
              style: const TextStyle(color: Colors.white),
            ),
          ],
        ),
      ),
    );
  }
}
```

### LINKED ISSUE #637: Screen freezes after scanner starts on Web using Flutter 3.10
I just migrated to Flutter 3.10 and I first struggle with [ISSUE 622](https://github.com/juliansteenbakker/mobile_scanner/issues/622). Then after adding JS dependency directly on index.html now camera starts successfully and QR scanning works but my screen widgets freezes, I mean I can tap on anything, no gestures are received. Widget tree consist in a Stack with `MobileScanner` at first and `BackButton` on top.

Using 
```yaml 
mobile_scanner: ^3.2.0
``` 
and setting 
```html
<script type="text/javascript" src="https://unpkg.com/@zxing/library@0.19.1"></script>
```
on index.html

Flutter doctor:
```
[✓] Flutter (Channel stable, 3.10.1, on macOS 13.0 22A380 darwin-arm64, locale en-US)
• Flutter version 3.10.1 on channel stable at /Users/user/fvm/versions/3.10.1
• Upstream repository https://github.com/flutter/flutter.git
• Framework revision d3d8effc68 (3 days ago), 2023-05-16 17:59:05 -0700
• Engine revision b4fb11214d
• Dart version 3.0.1
• DevTools version 2.23.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
• Android SDK at /Users/user/Android/SDK
• Platform android-33, build-tools 33.0.1
• ANDROID_HOME = /Users/user/Android/SDK
• Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
• Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
• All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
• Xcode at /Applications/Xcode.app/Contents/Developer
• Build 14E222b
• CocoaPods version 1.12.0

[✓] Android Studio (version 2022.1)
• Android Studio at /Applications/Android Studio.app/Contents
• Flutter plugin can be installed from:
🔨 https://plugins.jetbrains.com/plugin/9212-flutter
• Dart plugin can be installed from:
🔨 https://plugins.jetbrains.com/plugin/6351-dart
• Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
```

Check the following reproducible code sample:
```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:mobile_scanner/mobile_scanner.dart';

class SampleApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      debugShowCheckedModeBanner: true,
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  const Home({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: ElevatedButton(
          child: const Text('Go to Scanner Screen'),
          onPressed: () {
            Navigator.push(context, MaterialPageRoute(builder: (context) => CameraScannerSample()));
          },
        ),
      ),
    );
  }
}

class CameraScannerSample extends StatefulWidget {
  @override
  State createState() => CameraScannerSampleState();
}

class CameraScannerSampleState extends State {
  final cameraController = MobileScannerController(
    facing: CameraFacing.back,
    torchEnabled: false,
    formats: [BarcodeFormat.qrCode],
  );
  final focusOverlaySize = 200.0;
  bool isScanning = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: Stack(
        children: [
          if(isScanning)
            MobileScanner(
              controller: cameraController,
              onDetect: onDetect,
              scanWindow: Rect.fromCenter(
                center: MediaQuery.of(context).size.center(Offset.zero),
                width: focusOverlaySize,
                height: focusOverlaySize,
              ),
            )
          else Center(
            child: ElevatedButton(
              style: ElevatedButton.styleFrom(backgroundColor: Colors.red),
              onPressed: () => setState(() => isScanning = true),
              child: const Text('Start Scanning'),
            ),
          ),
          const Positioned(
            left: 16, top: 16,
            child: BackButton(
              color: Colors.red,
            ),
          ),
        ],
      ),
    );
  }

  void onDetect(BarcodeCapture barcodes) {
      if (kDebugMode) {
        print('---onNewScannerDecoded---');
      }
  }
}
```

### LINKED ISSUE #676: Uncaught zxing exception `TypeError: Cannot set properties of null (setting 'AbstractExpandedDecoder')`
I'm having a hard time tracking down what's happening here, but when starting up my app in VSCode under the debugger for the web target, I'm getting an uncaught exception:
`TypeError: Cannot set properties of null (setting 'AbstractExpandedDecoder')`

It doesn't give me any context about where this is raised from, and it's not triggering my `PlatformDispatcher.instance.onError` callback. Stepping through the code, it seems to be originating the first time `main` does an `await`. E.g. when I step through this code, it happens while stepping over the `await`:

```dart
Future<void> main() async {
  PlatformDispatcher.instance.onError = (error, stack) {
    FlutterError.presentError(FlutterErrorDetails(exception: error, stack: stack));
    return true;
  };

  await Future.delayed(const Duration(seconds: 1));
  WidgetsFlutterBinding.ensureInitialized();
```

image

This seems to be caught by the "Uncaught Exceptions" breakpoint in VS Code -- disabling it just means I get the following in the log:

```
TypeError: Cannot set properties of null (setting 'AbstractExpandedDecoder')
https://unpkg.com/@zxing/library@0.19.1 15:288561  <fn>
https://unpkg.com/@zxing/library@0.19.1 1:69       <fn>
https://unpkg.com/@zxing/library@0.19.1 1:206      <fn>
```

If I do that, I can also catch it in the Chrome debugger, and if I apply the source map, it turns into this line: `exports.AbstractExpandedDecoder = AbstractExpandedDecoder;` -- where `exports` is `null`.

Version info:
```
Flutter 3.10.5 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 796c8ef792 (12 days ago) • 2023-06-13 15:51:02 -0700
Engine • revision 45f6e00911
Tools • Dart 3.0.5 • DevTools 2.23.1
```

Using mobile_scanner from master (with https://github.com/juliansteenbakker/mobile_scanner/pull/675)

### PR DESCRIPTION
**Pull request by navaronbracke**
This PR adds the following changes:
- fix a bug in the static interop binding of `PhotoCapabilities`
- add the corners from the ZXing result to the barcode on web
- update the example app web entrypoint to the latest template by running `flutter create . --platforms=web`
- add better handling for the case where scanning barcodes is unsupported (for example a desktop running the browser sample)
- fix the permission denied handling on the web, by using the `NotAllowedError` error message as defined by MDN
- add app bars with back buttons to the example app (so that you can go back easily)

Fixes #637 
Fixes #622 (cannot reproduce the ZXing bug when manually adding the library in the body of the index.html)
Fixes #676
Fixes #632
Fixes #631 which is an umbrella issues for two of the issues in this list

### CODE REVIEW

File: lib/src/web/base.dart [hunk 1]
```diff
@@ -162,9 +160,16 @@ class Promise<T> {}
 
 @JS()
 @anonymous
-class PhotoCapabilities {
+@staticInterop
+class PhotoCapabilities {}
```
Comment (navaronbracke) [inline lib/src/web/base.dart:164]: This was the static interop bug

File: lib/src/objects/barcode.dart [hunk 2]
```diff
@@ -92,7 +92,8 @@ class Barcode {
 
   /// Create a [Barcode] from native data.
   Barcode.fromNative(Map data)
-      : corners = toCorners(data['corners'] as List?),
+      : corners =
```
Comment (navaronbracke) [inline lib/src/objects/barcode.dart:95]: I fixed all the unsafe casts here by fixing the types that are passed to `List.unmodifiable`

File: lib/src/mobile_scanner_controller.dart [hunk 3]
```diff
@@ -207,9 +207,21 @@ class MobileScannerController {
     } on PlatformException catch (error) {
       MobileScannerErrorCode errorCode = MobileScannerErrorCode.genericError;
 
-      if (error.code == "MobileScannerWeb") {
-        errorCode = MobileScannerErrorCode.permissionDenied;
+      final String? errorMessage = error.message;
+
+      if (kIsWeb) {
```
Comment (navaronbracke) [inline lib/src/mobile_scanner_controller.dart:212]: Using `kIsWeb` is safer, as the compiler provides it. I also updated the checks for unsupported / permission denied, per what I found on MDN.

File: lib/mobile_scanner_web_plugin.dart [hunk 4]
```diff
@@ -61,6 +61,8 @@ class MobileScannerWebPlugin {
         return _torch(call.arguments);
       case 'stop':
         return cancel();
+      case 'updateScanWindow':
```
Comment (navaronbracke) [inline lib/mobile_scanner_web_plugin.dart:64]: This being unimplemented resulted in an error in the console

**Comment by foxmind1** [comment]
I get this exception whenever i scan a QR code

`E/flutter ( 7390): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: type '_Map<Object?, Object?>' is not a subtype of type 'Map<String, Object?>' in type cast
E/flutter ( 7390): #0      _CastListBase.[] (dart:_internal/cast.dart:99:46)
E/flutter ( 7390): #1      ListBase.elementAt (dart:collection/list.dart:43:33)
E/flutter ( 7390): #2      MappedListIterable.elementAt (dart:_internal/iterable.dart:415:40)
E/flutter ( 7390): #3      ListIterator.moveNext (dart:_internal/iterable.dart:344:26)
E/flutter ( 7390): #4      new List.from (dart:core-patch/array_patch.dart:37:17)
E/flutter ( 7390): #5      new List.unmodifiable (dart:core-patch/array_patch.dart:67:24)
E/flutter ( 7390): #6      toCorners (package:mobile_scanner/src/barcode_utility.dart:17:15)
E/flutter ( 7390): #7      new Barcode.fromNative (package:mobile_scanner/src/objects/barcode.dart:96:13)
E/flutter ( 7390): #8      MobileScannerController._handleEvent.<anonymous closure> (package:mobile_scanner/src/mobile_scanner_controller.dart:368:37)
E/flutter ( 7390): #9      MappedListIterable.elementAt (dart:_internal/iterable.dart:415:31)
E/flutter ( 7390): #10     ListIterator.moveNext (dart:_internal/iterable.dart:344:26)
E/flutter ( 7390): #11     new _GrowableList._ofEfficientLengthIterable (dart:core-patch/growable_array.dart:189:27)
E/flutter ( 7390): #12     new _GrowableList.of (dart:core-patch/growable_array.dart:150:28)
E/flutter ( 7390): #13     new List.of (dart:core-patch/array_patch.dart:47:28)
E/flutter ( 7390): #14     ListIterable.toList (dart:_internal/iterable.dart:214:7)`

**Comment by navaronbracke** [comment]
Ah, I think I missed a cast somewhere then. I'll fix it up for you.

**Comment by foxmind1** [comment]
Thank you,
I will try when i come back home.

**Comment by foxmind1** [comment]
ok i don't get the exception anymore but it still doesn't work on web for me

**Comment by navaronbracke** [comment]
@foxmind1 It is working for me on the web. I can see the format, display value (might be null sometimes) and raw value.

You do have to use the CanvasKit renderer though, as drawing paths is not supported on the HTML renderer (for the scan window example)

File: lib/src/web/zxing.dart [hunk 5]
```diff
@@ -28,15 +39,24 @@ abstract class Result {
 
   /// Representing the format of the barcode that was decoded
   external int? format;
+
+  /// Returns the result points of the barcode. These points represent the corners of the barcode.
+  external List<Object?> get resultPoints;
```
Comment (navaronbracke) [inline lib/src/web/zxing.dart:44]: Since JS lists don't have a generic type, this is `Object?`

**Comment by navaronbracke** [comment]
@juliansteenbakker or @teolemon Friendly ping for getting this reviewed, as it solves a bunch of reported issues.

**Comment by foxmind1** [comment]
> 

ok so it actually works but for some reason it takes > 1 min to detect the QR code

**Comment by foxmind1** [comment]
> @juliansteenbakker or @teolemon Friendly ping for getting this reviewed, as it solves a bunch of reported issues.


ok so it actually works but for some reason it takes > 1 min to detect the QR code

**Comment by navaronbracke** [comment]
@foxmind1 I didn't notice any delay when scanning QR codes. For testing, I added some logging to the barcode capture event stream, which did log instantly upon scanning a QR code.

**Comment by foxmind1** [comment]
@navaronbracke yeah I'm not sure but for me sometimes it detect the code immediately and sometimes it takes more than a minute,
i found this issue https://github.com/juliansteenbakker/mobile_scanner/issues/136 it might be related.

**Comment by UliPrantz** [comment]
Unfortunately it doesn't seem to be fixed completely. On my machine I get the following error provided to the error widget builder when on web:
```
MobileScannerException: code genericError, message: TypeError: Cannot read properties of undefined (reading 'BrowserMultiFormatReader')
```
Tested on ```Chrome - Version 114.0.5735.198 (Official Build) (arm64)```

**Comment by navaronbracke** [comment]
@UliPrantz I do not get that exception with this fix. Bear in mind you have to include ZXing in your index.html (see the example app for an example). Also, this fix is only available on the master branch and has not been published yet.

Could you try with a git dependency to verify?

```
dependencies:
  mobile_scanner:
    git: git@github.com:juliansteenbakker/mobile_scanner.git
```

@juliansteenbakker I think you'd need to publish the fix :-)

**Comment by UliPrantz** [comment]
I was on the master as you said but yea you were right I completely forgot to add:
```
<body>
    // Add the following line
      <script src="https://unpkg.com/@zxing/library@0.19.1" type="application/javascript"></script>
    // Thats it....
    // other stuff and auto stuff generated starts here
</body>
```
This fixed it for me thank you. Definitely should mentioned later in the `README.md` once the fix is published.

**Comment by navaronbracke** [comment]
> I was on the master as you said but yea you were right I completely forgot to add:
> 
> ```
> <body>
>     // Add the following line
>       <script src="https://unpkg.com/@zxing/library@0.19.1" type="application/javascript"></script>
>     // Thats it....
>     // other stuff and auto stuff generated starts here
> </body>
> ```
>
> This fixed it for me thank you. Definitely should mentioned later in the `README.md` once the fix is published.

It is mentioned here https://github.com/juliansteenbakker/mobile_scanner#web but maybe we should surround it with the body tag like you did here.

**Pull request merged**